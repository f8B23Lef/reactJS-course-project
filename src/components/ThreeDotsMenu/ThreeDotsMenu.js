import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEllipsisV, faTimes } from '@fortawesome/free-solid-svg-icons';
import { useKeyDown } from '../../hooks/index';
import './ThreeDotsMenu.scss';

export default function ThreeDotsMenu(props) {
  const {
    isHovered,
    tdmItems,
  } = props;

  const [showTDM, setShowTDM] = useState(false);

  useKeyDown('Escape', () => setShowTDM(false));

  return (
    <div onClick={(event) => event.stopPropagation()}>
      <FontAwesomeIcon
        icon={faEllipsisV}
        className={`tdm__icon ${isHovered ? '' : 'hidden'}`}
        onClick={() => setShowTDM(true)}
      />
      <div className={`three-dots-menu ${showTDM ? '' : 'hidden'}`}>
        <FontAwesomeIcon
          icon={faTimes}
          className='tdm-close__icon'
          onClick={() => setShowTDM(false)}
        />
        {tdmItems.map(({ itemText, onClick }) => (
          <button
            type='button'
            key={itemText}
            onClick={() => { setShowTDM(false); onClick(); }}
          >
            {itemText}
          </button>
        ))}
      </div>
    </div>
  );
}

ThreeDotsMenu.propTypes = {
  isHovered: PropTypes.bool.isRequired,
  tdmItems: PropTypes.arrayOf(PropTypes.shape({
    itemText: PropTypes.string,
    onClick: PropTypes.func,
  })).isRequired,
};
