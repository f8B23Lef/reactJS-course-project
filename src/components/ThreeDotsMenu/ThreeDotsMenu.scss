@import '../../styles/index.scss';

.tdm__icon {
  display: block;
  position: absolute;
  top: 5px;
  right: 8px;
  width: 22px;
  height: 22px;
  border: 1px solid $dark-grey;
  border-radius: 50%;
  background-color: $dark-grey;
  cursor: pointer;

  &.hidden {
    display: none;
  }

  &:hover {
    background-color: $light-grey;
    border-color: $light-grey;
  }
}

.three-dots-menu {
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 5px;
  right: 5px;
  width: 100px;
  background-color: $dark-grey;
  border-radius: 2px;
  z-index: 2;

  &.hidden {
    display: none;
  }

  button {
    padding: 5px;
    border: none;
    text-align: start;
    color: $white;
    background-color: $dark-grey;
    cursor: pointer;

    &:hover {
      background-color: $grey;
    }
  }

  .tdm-close__icon {
    align-self: flex-end;
    padding: 5px;
    cursor: pointer;

    &:hover {
      color: $red;
    }
  }
}
