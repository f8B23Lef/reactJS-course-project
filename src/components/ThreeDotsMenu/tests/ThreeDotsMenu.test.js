import React, { useState as useStateMock } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import ThreeDotsMenu from '../ThreeDotsMenu';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

describe('ThreeDotsMenu component', () => {
  const onClickMock = jest.fn();
  const onDeleteMock = jest.fn();
  const tdmItems = [{
    itemText: 'Edit',
    onClick: onClickMock,
  }, {
    itemText: 'Delete',
    onClick: onDeleteMock,
  }];
  const isHovered = true;
  const setStateMock = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation((init) => [init, setStateMock]);
  });

  it('renders ThreeDotsMenu component snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <ThreeDotsMenu
          tdmItems={tdmItems}
          isHovered={isHovered}
        />
      </BrowserRouter>,
    );

    expect(asFragment(<ThreeDotsMenu
      tdmItems={tdmItems}
      isHovered={isHovered}
    />)).toMatchSnapshot();
  });

  tdmItems.forEach(({ itemText, onClick }) => {
    it(`calls onClick() from tdmItems prop on ${itemText} button click`, () => {
      render(
        <BrowserRouter>
          <ThreeDotsMenu
            tdmItems={tdmItems}
            isHovered={isHovered}
          />
        </BrowserRouter>,
      );

      const tdmButton = screen.getByText(itemText);
      fireEvent.click(tdmButton);

      expect(onClick).toHaveBeenCalled();
    });
  });

  it('calls setShowTDM() with true param on tdm icon click', () => {
    render(
      <BrowserRouter>
        <ThreeDotsMenu
          tdmItems={tdmItems}
          isHovered={isHovered}
        />
      </BrowserRouter>,
    );

    const icon = screen.getByTestId('tdm-icon');
    fireEvent.click(icon);

    expect(setStateMock).toHaveBeenCalledWith(true);
  });

  it('calls setShowTDM() with false param on close icon click', () => {
    render(
      <BrowserRouter>
        <ThreeDotsMenu
          tdmItems={tdmItems}
          isHovered={isHovered}
        />
      </BrowserRouter>,
    );

    const closeIcon = screen.getByTestId('tdm-close-icon');
    fireEvent.click(closeIcon);

    expect(setStateMock).toHaveBeenCalledWith(false);
  });
});
