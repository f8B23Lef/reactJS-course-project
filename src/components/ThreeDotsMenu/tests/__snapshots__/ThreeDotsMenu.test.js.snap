// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeDotsMenu component renders ThreeDotsMenu component snapshot 1`] = `
<DocumentFragment>
  <div>
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-ellipsis-vertical tdm__icon "
      data-icon="ellipsis-vertical"
      data-prefix="fas"
      data-testid="tdm-icon"
      focusable="false"
      role="img"
      viewBox="0 0 128 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M56 472a56 56 0 1 1 0-112 56 56 0 1 1 0 112zm0-160a56 56 0 1 1 0-112 56 56 0 1 1 0 112zM0 96a56 56 0 1 1 112 0A56 56 0 1 1 0 96z"
        fill="currentColor"
      />
    </svg>
    <div
      class="three-dots-menu hidden"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark tdm-close__icon"
        data-icon="xmark"
        data-prefix="fas"
        data-testid="tdm-close-icon"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M310.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L160 210.7 54.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L114.7 256 9.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 301.3 265.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L205.3 256 310.6 150.6z"
          fill="currentColor"
        />
      </svg>
      <button
        type="button"
      >
        Edit
      </button>
      <button
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
